File: cpp/src/parallel_quicksort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <thread>
#include <chrono>
#include <iomanip>

const size_t ARRAY_SIZE = 10'000'000;
const size_t THRESHOLD = 10'000;

size_t partition(std::vector<int>& arr, size_t low, size_t high) {
    int pivot = arr[high];
    size_t i = low;
    
    for (size_t j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            std::swap(arr[i], arr[j]);
            i++;
        }
    }
    
    std::swap(arr[i], arr[high]);
    return i;
}

void parallel_quicksort(std::vector<int>& arr, size_t low, size_t high) {
    if (low < high) {
        size_t len = high - low + 1;
        
        if (len <= THRESHOLD) {
            std::sort(arr.begin() + low, arr.begin() + high + 1);
            return;
        }
        
        size_t pi = partition(arr, low, high);
        
        if (pi > 0) {
            std::thread left_thread(parallel_quicksort, std::ref(arr), low, pi - 1);
            if (pi + 1 <= high) {
                parallel_quicksort(arr, pi + 1, high);
            }
            left_thread.join();
        } else {
            if (pi + 1 <= high) {
                parallel_quicksort(arr, pi + 1, high);
            }
        }
    }
}

int main() {
    // Generate data
    std::vector<int> data(ARRAY_SIZE);
    for (size_t i = 0; i < ARRAY_SIZE; i++) {
        data[i] = static_cast<int>((i * 1103515245 + 12345) % 2147483648);
    }
    
    // Warm-up
    std::vector<int> warmup = data;
    parallel_quicksort(warmup, 0, warmup.size() - 1);
    
    // Benchmark
    auto start = std::chrono::high_resolution_clock::now();
    parallel_quicksort(data, 0, data.size() - 1);
    auto end = std::chrono::high_resolution_clock::now();
    
    std::chrono::duration<double> duration = end - start;
    
    // Verify sort
    bool is_sorted = std::is_sorted(data.begin(), data.end());
    
    std::cout << std::fixed << std::setprecision(6) << duration.count() << std::endl;
    std::cerr << "Sorted: " << std::boolalpha << is_sorted << std::endl;
    
    return 0;
}

File: rust/src/parallel_quicksort.rs
use rayon;
use std::time::Instant;

const ARRAY_SIZE: usize = 10_000_000;
const THRESHOLD: usize = 10_000;

fn parallel_quicksort(arr: &mut [i32]) {
    if arr.len() <= THRESHOLD {
        arr.sort_unstable();
        return;
    }
    
    if arr.len() <= 1 {
        return;
    }
    
    let pivot_idx = partition(arr);
    let (left, right) = arr.split_at_mut(pivot_idx);
    
    rayon::join(
        || parallel_quicksort(left),
        || parallel_quicksort(&mut right[1..])
    );
}

fn partition(arr: &mut [i32]) -> usize {
    let len = arr.len();
    let pivot = arr[len - 1];
    let mut i = 0;
    
    for j in 0..len - 1 {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    
    arr.swap(i, len - 1);
    i
}

fn main() {
    // Generate random data
    let mut data: Vec<i32> = (0..ARRAY_SIZE)
        .map(|i| ((i * 1103515245 + 12345) % 2147483648) as i32)
        .collect();
    
    // Warm-up
    let mut warmup = data.clone();
    parallel_quicksort(&mut warmup);
    
    // Benchmark
    let start = Instant::now();
    parallel_quicksort(&mut data);
    let duration = start.elapsed();
    
    // Verify sort
    let is_sorted = data.windows(2).all(|w| w[0] <= w[1]);
    
    println!("{:.6}", duration.as_secs_f64());
    eprintln!("Sorted: {}", is_sorted);
}

